pkg/controller/dnsendpoint/dnsendpoint_controller.go

```diff
@@ -0,0 +1,217 @@
+	if len(managedDomains) == 0 {
+		logger.Info("no managed domains for external DNS; controller disabled")
+		return nil
+	}
+
+	nameServerQuery := nameserver.NewAWSQuery(c, credsSecretName)
+
+	nameServerChanges := make(chan event.GenericEvent, 1024)
+
+	registerNameServerChange := func(objectKey client.ObjectKey) {
+		nameServerChanges <- event.GenericEvent{
+			Meta: &metav1.ObjectMeta{
+				Namespace: objectKey.Namespace,
+				Name:      objectKey.Name,
+			},
+		}
+	}
+	nameServerScraper := newNameServerScraper(logger, nameServerQuery, managedDomains, registerNameServerChange)
+	if err := mgr.Add(nameServerScraper); err != nil {
+		return err
+	}
+
+	reconciler := &ReconcileDNSEndpoint{
+		Client:            c,
+		scheme:            mgr.GetScheme(),
+		logger:            logger,
+		nameServerScraper: nameServerScraper,
+		nameServerQuery:   nameServerQuery,
+	}
+	ctrl, err := controller.New(
+		controllerName,
+		mgr,
+		controller.Options{
+			Reconciler:              reconciler,
+			MaxConcurrentReconciles: controllerutils.GetConcurrentReconciles(),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	if err := ctrl.Watch(&source.Kind{Type: &hivev1.DNSEndpoint{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+	return ctrl.Watch(&source.Channel{Source: nameServerChanges}, &handler.EnqueueRequestForObject{})
+}
+
+var _ reconcile.Reconciler = &ReconcileDNSEndpoint{}
+
+// ReconcileDNSEndpoint reconciles a DNSEndpoint object
+type ReconcileDNSEndpoint struct {
+	client.Client
+	scheme            *runtime.Scheme
+	logger            log.FieldLogger
+	nameServerScraper *nameServerScraper
+	nameServerQuery   nameserver.Query
+}
+
+// Reconcile reads that state of the cluster for a DNSEndpoint object and makes changes based on the state read
+// and what is in the DNSEndpoint.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write DNSEndpoints
+func (r *ReconcileDNSEndpoint) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	dnsLog := r.logger.WithFields(log.Fields{
+		"controller":  controllerName,
+		"dnsendpoint": request.Name,
+		"namespace":   request.Namespace,
+	})
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	dnsLog.Info("reconciling dns endpoint")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+		dnsLog.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the DNSEndpoint object
+	instance := &hivev1.DNSEndpoint{}
+	if err := r.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		dnsLog.WithError(err).Error("Error fetching dnsendpoint object")
+		return reconcile.Result{}, err
+	}
+
+	if !isValidDNSEndpoint(instance, dnsLog) {
+		return reconcile.Result{}, nil
```

Comment:
For this step of changes, I did not want to make any changes to DNSEndpoint to avoid conflicts with existing clusterdeployments. The DNSEndpoint status does not have a place to set conditions. With future changes, I would like to remove the DNSEndpoint CRD entirely and have everything regarding managed DNS done entirely with DNSZone CRs. At a minimum, I would like to remove much of the generic DNS record cruft in DNSEndpoint, at which point this check would no longer be needed.